### PR TITLE
Celery_metrics

### PR DESCRIPTION
**Pull request by majamassarini**
Use flower service to collect celery related metrics (https://issues.redhat.com/browse/PACKIT-1109).

Related with packit/packit-service#1561

### CODE REVIEW

**Review by TomasTomecek — APPROVED** [review]

**Comment by TomasTomecek** [comment]
Where do we install the tool that provides the prometheus metrics? Or is this supported out of the box by flower?

File: openshift/nginx.yml.j2 [hunk 1]
```diff
@@ -67,9 +72,24 @@ data:
                 proxy_method     GET;
              }
 
+             location = /flower-metrics {
+               proxy_pass http://flower:5555/metrics;
+               proxy_method     GET;
+             }
+
              location / {
```
Comment (csomh) [inline openshift/nginx.yml.j2:80]:
```suggestion
             location /flower {
```

I think this should be rather something like this, in order to be explicit about where this route is going to lead :)

File: openshift/nginx.yml.j2 [hunk 2]
```diff
@@ -67,9 +72,24 @@ data:
...
-                return           404;
+               proxy_pass http://flower:5555;
+               proxy_set_header Host $host;
+               proxy_set_header X-Real-IP $remote_addr;
+
+               proxy_http_version 1.1;
+               proxy_set_header Upgrade $http_upgrade;
+               proxy_set_header Connection "upgrade";
```
Comment (csomh) [inline openshift/nginx.yml.j2:87]: Why are these required? Maybe add a comment or link explaining it.

File: playbooks/deploy.yml [hunk 3]
```diff
@@ -359,13 +360,47 @@
...
+
+    - name: Create htpasswd file with flower-boss passwd
+      htpasswd:
+        path: "{{ flower_secret_htpasswd }}"
+        name: "flower-boss"
+        password: "{{ flower_basic_auth | regex_replace('flower-boss:', '') }}"
+        mode: 0640
+      tags:
+        - flower
+      when: with_flower
```
Comment (csomh) [inline playbooks/deploy.yml:383]: Seeing it like this, I think it would be much easier to not to generate this file on every Ansible playbook run, but to generate it manually and save it as an attachment in Bitwarden. Then it could be deployed with the rest of the secrets.

Comment (csomh) [reply playbooks/deploy.yml:383]: And we wouldn't need to introduce new dependencies :)

File: openshift/nginx.yml.j2 (hunk 1 above)
Comment (csomh) [reply openshift/nginx.yml.j2:80]: And keep returning `404` on `/`.

**Comment by majamassarini** [comment]
Flower has already all the metrics provided by celery-exporter, for this reason we decided (with Laura) to use it.
This changes are needed to make the flower metrics be reachable (by prometheus) without any authentication and let all the other flower services be under authentication as they already were.

File: openshift/nginx.yml.j2 (hunk 1 above)
Comment (majamassarini) [reply openshift/nginx.yml.j2:80]: I tried really hard to do that, without any success.
I was not able to serve the static content as expected by flower and at the end I give up... if you think this is necessary we can try to configure it together.

File: playbooks/deploy.yml (hunk 3 above)
Comment (majamassarini) [reply playbooks/deploy.yml:383]: We need both the plain password (to access flower services manually) and the crypted password (to generate the htpasswd like file for nginx).
If we do it by hand we should cope with the "duplication" of having the same password in two different formats.
I am not really sure that this is a good idea but, on the other hand, I am not sure that my approach is the easiest and cleanest, maybe do we have some other way to do that?

File: openshift/nginx.yml.j2 (hunk 1 above)
Comment (csomh) [reply openshift/nginx.yml.j2:80]: After playing with this a little bit, I don't think this can be done, or at least I wasn't able to figure it out either. Flower needs some more resources (CSS, JS) loaded to run, and loading these resources will happen using the hostname, but in order to work, it should use `hostname/flower`. And I wasn't able to figure out how to configure this in flower.

So let's keep everything as it is.

File: playbooks/deploy.yml (hunk 3 above)
Comment (csomh) [reply playbooks/deploy.yml:383]: I don't mind the duplication in this case, as it will help reduce complexity.

File: playbooks/deploy.yml [hunk 4]
```diff
@@ -359,13 +360,47 @@
...
+        path: "{{ flower_secret_htpasswd }}"
+        state: absent
+      tags:
+        - flower
+      when: with_flower
+
+    - name: Show flower-boss passwd
+      ansible.builtin.debug:
+        msg: "{{ flower_basic_auth | regex_replace('flower-boss:', '') }}"
+
```
Comment (csomh) [inline playbooks/deploy.yml:374]:
```suggestion
```

Could be removed.

File: playbooks/deploy.yml [hunk 5]
```diff
@@ -359,13 +360,47 @@
...
+    - name: Create htpasswd file with flower-boss passwd
+      htpasswd:
```
Comment (csomh) [inline playbooks/deploy.yml:376]:
```suggestion
      community.general.htpasswd:
```

File: playbooks/deploy.yml [hunk 6]
```diff
@@ -359,13 +360,47 @@
...
 
-    - name: Deploy nginx to reverse proxy the pushgateway
+    - name: Delete previous created file with flower credentials for nginx, if any
+      ansible.builtin.file:
+        path: "{{ flower_secret_htpasswd }}"
+        state: absent
+      tags:
+        - flower
+      when: with_flower
+
```
Comment (csomh) [inline playbooks/deploy.yml:370]: There is no need to do this, as the `htpasswd` module will update the file if any of the credentials do change.

File: playbooks/deploy.yml (hunk 3 above)
Comment (jpopelka) [reply playbooks/deploy.yml]: Both solutions have pros and cons. First I leaned toward Hunor's suggestion to have "duplicated" file(s) in the vault rather than more code here in the playbook. But then, we would need to document somewhere how we actually created that file, because I definitely won't remember after some time. So now I'm +1 for creating the file on the fly as it's just a few simple lines here in the playbook.

File: vars/packit/prod_template.yml [hunk 7]
```diff
@@ -31,7 +31,7 @@ api_key: ""
 
 # with_redis_commander: false
 
-# with_flower: false
+with_flower: true
```
[the remark this answers is not in the document]
Comment (jpopelka) [reply vars/packit/prod_template.yml:34]: forget about it :-) the main template.yml is fine since we're not changing the default value of `with_flower`

File: playbooks/deploy.yml [hunk 8]
```diff
@@ -28,6 +28,7 @@
     image_tokman: quay.io/packit/tokman:{{ deployment }}
     # project_dir is set in tasks/project-dir.yml
     path_to_secrets: "{{ project_dir }}/secrets/{{ service }}/{{ deployment }}"
+    flower_secret_htpasswd: "{{ path_to_secrets }}/flower-htpasswd"
```
[the remark this answers is not in the document]
Comment (majamassarini) [reply playbooks/deploy.yml:31]: Nice, thanks for explaining me this!

**Review by csomh — APPROVED** [review]
